csrc/multidevice/executor.cpp

```diff
@@ -155,12 +155,13 @@ void MultiDeviceExecutor::postKernel(SegmentedGroup* group) {
 
   // Compile the group and execute it with FusionExecutor
   // Check if the executor has been cached. If not, create and cache it
-  if (fe_.find(group) == fe_.end()) {
-    fe_.emplace(group, std::make_unique<FusionExecutor>());
-    fusions_.emplace(group, staged_fusion_->makeFusion(group));
-    fe_[group]->compileFusion(fusions_.at(group).get(), group_input_IValues);
+  if (fec_.find(group) == fec_.end()) {
+    fec_.emplace(
+        group,
+        std::make_unique<FusionExecutorCache>(
+            staged_fusion_->makeFusion(group), 0, auto_schedule_));
```

Comment:
I doubt the use of unique_ptr is necessary, which adds execution overhead and indirection for readers. I *think* something like 
```suggestion
    fec_.try_emplace(
        group,
        staged_fusion_->makeFusion(group), 0, auto_schedule_);
```
should just work. (Of course, you'd need to change fec_'s definition as well)

Reply:
I agree with you on principle, and this is what I tried initially, but the compiler complains:

```
FAILED: CMakeFiles/codegen_internal.dir/csrc/multidevice/executor.cpp.o
/usr/bin/c++ -DNVFUSER_DISTRIBUTED -DTORCH_CUDA_BUILD_MAIN_LIB -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_C10D_NCCL -DUSE_C10D_UCC -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -I/opt/pytorch/Fuser/csrc -I/home/snordmann/nvfuser_build/include -I/opt/pytorch/Fuser/lib/dynamic_type/src -isystem /opt/pytorch/Fuser/third_party/gloo -isystem /opt/pytorch/Fuser/third_party/flatbuffers/include -isystem /usr/local/cuda/extras/CUPTI/include -isystem /usr/local/cuda/include -isystem /opt/p
ytorch/pytorch/torch/include -isystem /opt/pytorch/pytorch/torch/include/torch/csrc/api/include -Wno-psabi -D_GLIBCXX_USE_CXX11_ABI=1 -g -std=gnu++17 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wno-unused-function -Werror -D_GLIBCXX_USE_CXX11_ABI=1 -MD -MT CMakeFiles/codegen_internal.dir/csrc/multidevice/executor.cpp.o -MF CMakeFiles/codegen_internal.dir/csrc/multidevice/executor.cpp.o.d -o CMakeFiles/codegen_internal.dir/csrc/multidevice/executor.cpp.o -c
 /opt/pytorch/Fuser/csrc/multidevice/executor.cpp
In file included from /usr/include/x86_64-linux-gnu/c++/11/bits/c++allocator.h:33,
                 from /usr/include/c++/11/bits/allocator.h:46,
                 from /usr/include/c++/11/unordered_map:40,
                 from /usr/include/c++/11/functional:61,                                                                                                                                                                                                                                                                                                                                                                                                                                                      from /opt/pytorch/pytorch/torch/include/c10/core/Allocator.h:5,
                 from /opt/pytorch/pytorch/torch/include/ATen/cuda/CUDAContextLight.h:24,
                 from /opt/pytorch/pytorch/torch/include/ATen/cuda/CUDAContext.h:3,
                 from /opt/pytorch/Fuser/csrc/multidevice/executor.cpp:9:
/usr/include/c++/11/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>; _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _Tp = std::__detail::_Hash_node<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, false>]’:
/usr/include/c++/11/bits/alloc_traits.h:516:17:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>; _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _Tp = std::__detail::_Hash_node<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, false
>; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<std::__detail::_Hash_node<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, false> >]’
/usr/include/c++/11/bits/hashtable_policy.h:1878:36:   required from ‘std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type* std::__detail::_Hashtable_alloc<_NodeAlloc>::_M_allocate_node(_Args&& ...) [with _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, false> >; std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_ptr = std::allocator<std::__detail::_Hash_node<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, false> >::value_type*]’
/usr/include/c++/11/bits/hashtable.h:304:35:   required from ‘std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::_Scoped_node::_Scoped_node(std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::__hashtable_alloc*, _Args&& ...) [with _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _Key = nvfuser::SegmentedGroup*; _Value = std::pair<nvfuse
r::SegmentedGroup* const, nvfuser::FusionExecutorCache>; _Alloc = std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<nvfuser::SegmentedGroup*>; _Hash = std::hash<nvfuser::SegmentedGroup*>; _RangeHash = std::__detail::_Mod_range_hashing; _Unused = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_t
raits<false, false, true>; std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::__hashtable_alloc = std::_Hashtable<nvfuser::SegmentedGroup*, std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >, std::__detail::_Select1st, std::equal_to<nvfuser::SegmentedGroup*>, std::hash<nvfuser::SegmentedGroup*>, std::__d
etail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::__hashtable_alloc]’
/usr/include/c++/11/bits/hashtable.h:1966:15:   required from ‘std::pair<typename std::__detail::_Insert<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::iterator, bool> std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::_M_emplace(std::true_type, _Args&& ...) [with _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _Key = nvfuser::SegmentedGroup*; _V
alue = std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>; _Alloc = std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<nvfuser::SegmentedGroup*>; _Hash = std::hash<nvfuser::SegmentedGroup*>; _RangeHash = std::__detail::_Mod_range_hashing; _Unused = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std:
:__detail::_Hashtable_traits<false, false, true>; typename std::__detail::_Insert<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::iterator = std::__detail::_Insert_base<nvfuser::SegmentedGroup*, std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >, std::__detail::_Select1st, std::equal_to<nvfuser::SegmentedGroup*>, std::hash<nvfuser::SegmentedGroup*>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::iterator; typename _Traits::__constant_iterators = std::__detail::_Hashtable_traits<false, false, true>::__constant_iterators; std::true_type = std::integral_constant<bool, true>]’
/usr/include/c++/11/bits/hashtable.h:915:21:   required from ‘std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::__ireturn_type std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::emplace(_Args&& ...) [with _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _Key = nvfuser::SegmentedGroup*; _Value = std::pair<nvfuser::SegmentedGroup* const
, nvfuser::FusionExecutorCache>; _Alloc = std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<nvfuser::SegmentedGroup*>; _Hash = std::hash<nvfuser::SegmentedGroup*>; _RangeHash = std::__detail::_Mod_range_hashing; _Unused = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<false, false, true
>; std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _Hash, _RangeHash, _Unused, _RehashPolicy, _Traits>::__ireturn_type = std::_Hashtable<nvfuser::SegmentedGroup*, std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >, std::__detail::_Select1st, std::equal_to<nvfuser::SegmentedGroup*>, std::hash<nvfuser::SegmentedGroup*>, std::__detail::_Mod_range_hashing,
std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::__ireturn_type]’
/usr/include/c++/11/bits/unordered_map.h:389:23:   required from ‘std::pair<typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__is_nothrow_invocable<const _Hash&, const _Key&> > >::value, false, true> >::iterator, bool>
std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::emplace(_Args&& ...) [with _Args = {nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache}; _Key = nvfuser::SegmentedGroup*; _Tp = nvfuser::FusionExecutorCache; _Hash = std::hash<nvfuser::SegmentedGroup*>; _Pred = std::equal_to<nvfuser::SegmentedGroup*>; _Alloc = std::allocator<std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >; typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, s
td::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__is_nothrow_invocable<const _Hash&, const _Key&> > >::value, false, true> >::iterator = std::__detail::_Insert_base<nvfuser::SegmentedGroup*, std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>, std::allocator<std::pa
ir<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache> >, std::__detail::_Select1st, std::equal_to<nvfuser::SegmentedGroup*>, std::hash<nvfuser::SegmentedGroup*>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true> >::iterator]’
/opt/pytorch/Fuser/csrc/multidevice/executor.cpp:159:17:   required from here
/usr/include/c++/11/ext/new_allocator.h:162:11: error: no matching function for call to ‘std::pair<nvfuser::SegmentedGroup* const, nvfuser::FusionExecutorCache>::pair(nvfuser::SegmentedGroup*&, nvfuser::FusionExecutorCache)’
  162 |         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
I believe you missed the `try_` part. Anyhow, I tried the following on **your** branch and it builds. 

```
diff --git a/csrc/multidevice/executor.cpp b/csrc/multidevice/executor.cpp
index 52120d2..e3e0226 100644
--- a/csrc/multidevice/executor.cpp
+++ b/csrc/multidevice/executor.cpp
@@ -155,13 +155,8 @@ void MultiDeviceExecutor::postKernel(SegmentedGroup* group) {

   // Compile the group and execute it with FusionExecutor
   // Check if the executor has been cached. If not, create and cache it
-  if (fec_.find(group) == fec_.end()) {
-    fec_.emplace(
-        group,
-        std::make_unique<FusionExecutorCache>(
-            staged_fusion_->makeFusion(group), 0, auto_schedule_));
-  }
-  outputs = fec_[group]->runFusionWithInputs(group_input_IValues);
+  fec_.try_emplace(group, staged_fusion_->makeFusion(group), 0, auto_schedule_);
+  outputs = fec_.at(group).runFusionWithInputs(group_input_IValues);

   // Store the outputs in the context
   for (auto output_idx : c10::irange(outputs.size())) {
diff --git a/csrc/multidevice/executor.h b/csrc/multidevice/executor.h
index 49fe0b7..fe25418 100644
--- a/csrc/multidevice/executor.h
+++ b/csrc/multidevice/executor.h
@@ -119,8 +119,7 @@ class MultiDeviceExecutor {
   // Stores the order in which the pipeline's stage should be executed
   RuntimeWorkSpace workspace;
   // Cache FusionsExecutorCache
-  std::unordered_map<SegmentedGroup*, std::unique_ptr<FusionExecutorCache>>
-      fec_;
+  std::unordered_map<SegmentedGroup*, FusionExecutorCache> fec_;
   // Cache whether a SegmentedGroup should be run by the current device
   std::unordered_map<SegmentedGroup*, bool> should_run_;
   // Cache whether a SegmentedGroup requires inter-device communication
```

Reply:
(I would have committed this directly if I had been added to your repo!)